### PR TITLE
[8.0][ADD] membership_withdrawal

### PR DESCRIPTION
**Pull request by antespi**
Membership withdrawal
=====================

This module allows to record membership withdrawal reason and date of request.

Configuration
=============

Users can define membership withdrawal reasons in Association > Configuration > Membership withdrawal reasons
Then go to membership lines and set:

* Withdrawal reason: Why member requested withdrawal
* Withdrawal date: When member requested withdrawal
* Date cancel: Date when membership period must finish

### CODE REVIEW

**Comment by antespi** [comment]
cc @Tecnativa

**Review by rafaelbn — APPROVED** [review]
Tested in runbot

File: membership_withdrawal/README.rst [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+Configuration
+=============
+
+Users can define membership withdrawal reasons in Association > Configuration > Membership withdrawal reasons
+Then go to membership lines and set:
+
+* Withdrawal reason: Why member requested withdrawal
+* Withdrawal date: When member requested withdrawal
+* Date cancel: Date when membership period must finish
```
Comment (pedrobaeza) [inline membership_withdrawal/README.rst:19]: Is "Date cancel" the best term? Maybe "Ending date" o "Theoretical end"

File: membership_withdrawal/data/membership_withdrawal_reason_data.xml [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+<openerp>
+<data>
+
+<record id="reason_unsatisfied" model="membership.withdrawal_reason">
+    <field name="name">Unsatisfied</field>
+</record>
+
+<record id="reason_company_bankrupt" model="membership.withdrawal_reason">
+    <field name="name">Company bankrupt</field>
+</record>
```
Comment (pedrobaeza) [inline membership_withdrawal/data/membership_withdrawal_reason_data.xml:13]: Provide a generic one "Other"

File: membership_withdrawal/models/membership_line.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+# Copyright 2016 Antonio Espinosa <antonio.espinosa@tecnativa.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from openerp import models, fields
+
+
+class MembershipLine(models.Model):
+    _inherit = "membership.membership_line"
+
+    date_withdrawal = fields.Date(
```
Comment (pedrobaeza) [inline membership_withdrawal/models/membership_line.py:11]: For coherence, you should rename it to `withdrawal_date`

File: membership_withdrawal/models/res_partner.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+            else:
+                withdrawal_reason_id = False
+                date_withdrawal = False
+                for line in partner.member_lines:
+                    if line.withdrawal_reason_id and line.date_withdrawal:
+                        withdrawal_reason_id = line.withdrawal_reason_id
+                        date_withdrawal = line.date_withdrawal
+                        break
+                partner.membership_last_withdrawal_reason_id = \
+                    withdrawal_reason_id.id if withdrawal_reason_id else False
```
Comment (pedrobaeza) [inline membership_withdrawal/models/res_partner.py:43]: You only need `withdrawal_reason_id.id`

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

File: membership_withdrawal/README.rst (hunk 1 above)
Comment (antespi) [reply membership_withdrawal/README.rst:19]: This field is a core one

File: membership_withdrawal/models/membership_line.py (hunk 3 above)
Comment (antespi) [reply membership_withdrawal/models/membership_line.py:11]: All dates fields in this model start with date_*

File: membership_withdrawal/models/res_partner.py (hunk 4 above)
Comment (antespi) [reply membership_withdrawal/models/res_partner.py:43]: withdrawal_reason_id = False in line 35, so if no member_lines then I have to check it

File: membership_withdrawal/data/membership_withdrawal_reason_data.xml (hunk 2 above)
Comment (antespi) [reply membership_withdrawal/data/membership_withdrawal_reason_data.xml:13]: Done

**Comment by antespi** [comment]
Rebased and squashed

**Comment by rafaelbn** [comment]
Hello, what's missing here? Functionally is working.

File: membership_withdrawal/models/res_partner.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from openerp import models, fields, api
+
+
+class ResPartner(models.Model):
+    _inherit = "res.partner"
+
+    membership_last_withdrawal_reason_id = fields.Many2one(
+        string="Membership withdrawal reason", store=True, readonly=True,
```
Comment (susport) [inline membership_withdrawal/models/res_partner.py:12]: It is not necessary readonly if you field is compute.

File: membership_withdrawal/models/membership_withdrawal_reason.py [hunk 6]
```diff
@@ -0,0 +1,12 @@
+# -*- coding: utf-8 -*-
+# Copyright 2016 Antonio Espinosa <antonio.espinosa@tecnativa.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from openerp import models, fields
```
Comment (susport) [inline membership_withdrawal/models/membership_withdrawal_reason.py:5]: Order by name the import according to the OCA guideline

**Comment by pedrobaeza** [comment]
Closing as we have moved target to 10.0 in #33

**Pull request closed without merging**